Fix embedding through views with subqueries inside function calls

wolfgangwalther opened the pull request:
Resolves #1608, resolves #1635.

This is on top of the changes in #1625. Will rebase, once the other one is merged.

Background:
Due to the nested nature of `pg_node_tree`, it is not possible to correctly parse all kinds of views with just a regex. This replaces that parsing by using the postgres `json` parser. First the `pg_node_tree` is reformatted to proper json with a series of `replace()` and `regexp_replace()` calls. After that, a cast to json is possible and the TargetEntries can easily be extracted.

The json that is produced is a very much stripped down version of the node tree, that has just two parts left:
- the fields that we require for our query
- the whole structure, i.e. all object and array brackets with some empty string keys, where required

It looks roughly like this:
```json
[
  {
    "": [
      {"":{},"":{}},
      {"":{},"":{}},
      {"":{}},
      {"":{}},
      {"":{}},
      {"":{},"":[{},{},{},{},{},{},{},{},{},{},{},{},{},{}]},
      {"":{},"":[{},{},{},{},{},{},{},{},{},{},{},{},{},{},{},{},{}]}
    ],
    "": {
      "":[{"":{},"":{"":{},"":{},"":{"":[{},{}]}},"":{"":[{},{}]}}],
      "":{"":[{},{"":[{},{},{}]}]}
    },
    "targetList": [
      {
        "": {},
        "resno": 1,
        "resorigtbl": 27118,
        "resorigcol": 1
      },
      {
        "": {},
        "resno": 2,
        "resorigtbl": 27118,
        "resorigcol": 2
      },
      {
        "": {},
        "resno": 3,
        "resorigtbl": 27118,
        "resorigcol": 3
      }
    ],
    "": [{},{}]
  }
]
```

Using the big test schema, I get 1 more available column to embed on and the following timings for n=100:

```
query  | plan [ms] | execution [ms]
master | 1.71      | 681.5
#1625  | 2.60      | 417.5
json   | 2.67      | 574.2
```

Those numbers are from a different (less powerful) machine compared to those reported in #1625. Here I don't get a factor of 2x between master and the optimized regex query, more like 1,5x. Considering that I started with ~8000ms for the json approach, I'm quite happy that I was able to make it faster than the current master.

While #1625 is even faster, the json approach has a lot of upside, not only the "subquery inside a function call" that I had fixed initially. Since we now have the full node tree available, we can extend the parsing to support set operations (unions, ...). This will be part of a diffrent PR, though. Once we support UNION, we should be able to give embedding hints (#1179) like this:

```sql
CREATE VIEW xy AS
SELECT a, b, foreign_key_column FROM base_table WHERE false
UNION
SELECT my_regular_view_which_is_not_detected_because_of_crazy_functions
FROM base_table
```

The first select would basically serve as an embedding hint producing no rows. 

_@steve-chavez: Since this is still faster than current master/v7, I think you will still be able to call it "The Performance Release" if you want to, even when you merge both #1625 and this one. ;)_

steve-chavez commented:
@wolfgangwalther You're on fire :fire:!

I was rushing to do #1633, but tomorrow I'll get back to reviewing #1625.

"The Performance Release" name will stay ;)

wolfgangwalther commented:
rebase on current version of #1625

wolfgangwalther commented:
rebased on master

wolfgangwalther commented:
just a rebase... to get rid of the red CI checks. not true! This one is good!

wolfgangwalther commented:
With some of the other PRs done now - @steve-chavez did you have a chance to look at this one already? Is there any kind of testing that we should still do in your opinion?

I have one point, that we need to think through:
I'm quite confident that the JSON parsing for pg_node_tree should work for all trees we will see. But if for whatever reason one of them fails, the whole query will fail. I don't know a way of casting to JSON with a silent `NULL` fallback if casting fails - short of creating a stored procedure with exception handling. But we don't have our `pgrst` schema, yet.

I'm not entirely sure, but I think that the schema query currently fails silently if an sql exception is raised. I don't think we get log output of that. What we should really do is add proper error handling to those schema queries - so that we get log output on the console when that happens. Since we never expect any errors in those queries, they could include a line similiar to "The following error occured... This is a bug in PostgREST. Please report this at githuburl with this error message".

steve-chavez commented on src/PostgREST/DbStructure.hs:
```diff
@@ -738,36 +731,100 @@ pfkSourceColumns cols pgVer =
+            -- pg_node_tree has ` ` between list items, but JSON uses `,`
+            ), '} {'             , '},{'
+            ), '] ['             , '],['
+            -- `<>` in pg_node_tree is the same as `null` in JSON, but due to very poor performance of json_typeof
```
> but due to very poor performance of json_typeof

@wolfgangwalther Oh, how bad is it? We're using json_typeof for our INSERTs/UPDATEs.

wolfgangwalther replied:
I remember this was really bad. I might be able to do some testing, to confirm that. With `json_typeof` the query looked a bit different - so maybe it wasn't directly json_typeof, but the other query structure in general.

wolfgangwalther replied:
So I just tested this for this query again. The changes I made where:
```
), '<>'              , '[]'
```
becomes
```
), '<>'              , 'null'
```
(so no target list is encoded as null instead of an empty array)

and then

```
target_entries as(
        select
          view_id, view_schema, view_name,
          json_array_elements(view_definition->0->'targetList') as entry
        from transform_json
        where json_typeof(view_definition->0->'targetList') = 'array' -- this line is added
      ),
```

so empty target lists are now filtered by json_typeof.

The changed replace should have no effect, so only json_typeof makes things slower here. Here are the results:

| query | execution |
| -- | --: |
| old master | 415.0 |
| master | 232.6 |
| json PR | 321.9 |
| json_typeof | 714.1 |

So just the json_typeof call takes more time then everything else combined - the time is more than doubled!

wolfgangwalther replied:
I tested a bit more and even without `json_typeof`, just adding the following make the query just as slow:

```
where view_definition::text is not NULL
```

This is effectively a no-op - but the cast to `text` is taking quite a bit of time. It probably has to go through the whole json object. The same thing will happen here for `json_typeof`.

We're currently using json_typeof here:
https://github.com/PostgREST/postgrest/blob/11d62a80105c21696e65b70acf07b6f47405ca68/src/PostgREST/Private/QueryFragment.hs#L40-L53

Maybe we can just check the first non-whitespace character of the body we're passing in, in haskell. If it's a `[` use the array code, if it's a `{` use the object code. Could give us a performance improvement. Or not. Would have to loadtest this, of course :)

steve-chavez replied:
Ahh.. I see it now! Maybe we could make the `[ or {` check  on pg?

I feel like messing with the payload on Haskell might bring https://github.com/PostgREST/postgrest/issues/690 again.

Or perhaps we should just accept the `json_typeof` overhead. It's nice to be aware of it though.

wolfgangwalther replied:
> Ahh.. I see it now! Maybe we could make the `[ or {` check on pg?

Yeah, I thought so. That was the reason why I tried `where view_definition::text is not NULL`. Because to do the check, we need to cast the json back to text first. Just as slow as `pg_typeof`.

> I feel like messing with the payload on Haskell might bring #690 again.

Yeah, we would need to make sure that we're not parsing the body at all. We just want to read the first bytes of it. Isn't that kind of lazy evaluation one of haskell's strength? If we do it properly of course.

That still does not mean the json is correct in any way. But in that case PG would still throw.

Plus, it would also solve #1323. I tried to debug that today, but I don't see why this is happening. This would be fixed by:
- if `[` -> recordset
- if `{` -> record
- else -> error

wolfgangwalther replied:
Still had this one my todo list - but I guess we can call this resolved. My tests were flawed, it wasn't really `json_typeof` that was performing that bad, it was the "accessing nested json values" part that came with the approach.

`json_typeof` does have a little bit of overhead - but that's to be expected when it's called ~2800 times when querying the schema cache in *the big schema*. I don't see a problem when it's called once per request.

steve-chavez commented on src/PostgREST/DbStructure.hs:
```diff
@@ -699,18 +699,11 @@ pkFromRow tabs (s, t, n) = PrimaryKey <$> table <*> pure n
 -- returns all the primary and foreign key columns which are referenced in views
-pfkSourceColumns :: [Column] -> PgVersion -> Bool -> H.Statement ([Schema], [Schema]) [SourceColumn]
-pfkSourceColumns cols pgVer =
+pfkSourceColumns :: [Column] -> Bool -> H.Statement ([Schema], [Schema]) [SourceColumn]
+pfkSourceColumns cols =
   H.Statement sql (contrazip2 (arrayParam HE.text) (arrayParam HE.text)) (decodeSourceColumns cols)
-  -- query explanation at https://gist.github.com/steve-chavez/7ee0e6590cddafb532e5f00c46275569
```
@wolfgangwalther I think this one will be hard to understand by just looking at the code. Even on my prior query I had to think hard to remember what I did.

Could you add a comment here explaining the general process?

> First the pg_node_tree is reformatted to proper json with a series of replace() and regexp_replace() calls. After that, a cast to json is possible and the TargetEntries can easily be extracted.

Even a comment like the above would be helpful.

If you have some sample node trees like on my prior gist, that would be helpful as well. You can also leave a gist of your own if you'd like to.

wolfgangwalther replied:
Yes that makes a lot of sense. I removed the link to your gist, because big parts of it don't apply anymore (the regex stuff). But the basics still do and you're absolutely right that this could use a bit more explanation here. I tried to comment the json conversion as much as possible - but that does not give any idea of the bigger picture here.

I will try to add more comments.

wolfgangwalther replied:
I forked your gist and took the regex stuff out. This is part #1. Then I wrote part #2 about the json transformation in detail.

First of all, this is lots of detail, very good to come back to that in the future. But even better - explaining it in detail, I realized that I could skip one of the replaces altogether. My timings are now:

| query | execution |
| -- | --: |
| old master | 406.5 |
| master | 229.8 |
| json before | 317.5 |
| json now | 301.8 |

wolfgangwalther replied:
You probably won't notice on your super fast machine, but on my slow Ryzen it's 16ms faster ;)

steve-chavez replied:
Awesome! That's a really nice explanation Wolfgang, I feel we have at least 2 blog posts with those gists :)

> You probably won't notice on your super fast machine, but on my slow Ryzen it's 16ms faster ;)

I'll try the new query to see if I can beat your time again :smile:. Do you have a quick pastebin for me? :)

wolfgangwalther replied:
https://pastebin.com/6x3wTJVX

Here you go :)

steve-chavez replied:
| query       | execution          |
| ----------- | ------------------ |
| old master  | 282.7              |
| master      | 168.6              |
| json before | 249.2              |
| json now    | 232.0              |

17 ms faster for me. Beat you again :laughing:

wolfgangwalther replied:
Ah, wait a second. Between my latest test and yours... I removed another replace. It didn't show much of a difference in perf, so I didn't post it... but your test was running with both of them removed..

So let me run my tests again... let's see whether I can get another 1 or 2 ms. :D

wolfgangwalther replied:
319.9-303.5 = 16.4. :(

I'm leaving now to do something else. This is annoying :D

steve-chavez commented:
> Using the big test schema, I get 1 more available column to embed on and the following timings for n=100:
query  | plan [ms] | execution [ms]
master | 1.71      | 681.5
json   | 2.67      | 574.2

Bummer, I'm seeing that the new query is about x2 slower than the old one. Same setup with the `timeit` function(n=100):

| query | plan(ms) | execution(ms) |
| ----- | -------- | ------------- |
| master   | 3.38227  | 618.11523     |
| json   | 3.52706  | 1450.29681    |

@wolfgangwalther Have you tested the query again recently? What could be wrong?

steve-chavez commented:
>  I get 1 more available column to embed 

I get 10 more records on the big schema when running the new query(578 on the old, and 588 on the new), perhaps #1625 changed the perf here?

wolfgangwalther commented:
I just did another performance test. This is what I did to set up the test:
- **old master**: Took query from commit 65968b532060bc59452e354c96888b85057e67b7 (just before merging #1625), replaced `$1` with `'{apflora}'`, replaced `{subselectRegex}` with the regex for "otherwise", so >= pg10, remember to change `\\d` to `\d` here.
- **current master**: took current master, replaced `$1 || $2` with `'{apflora,public}'` and further down `$1` with `'{apflora}'`, same as for old master regarding the regex
- **json PR**: took this PR, replaced the parameters the same as on current master, no replacement of regex, because this is not there anymore

You can find the whole SQL here: https://pastebin.com/9CYCzCxc

Here are my results (n=100) - running on PG 12.5:

| query | plan | execution | count |
| -- | --: | --: | --: |
| old master | 0.72 | 460.0 | 2243 |
| master | 1.40 | 258.7 | 587 |
| json PR | 1.34 | 337.3 | 588 |

This matches my expectations:
1. **number of rows**: old master returned all columns, we changed that to only return pk and fk columns in #1625. In the big schema there is only 1 new source column detected, this must be because of some subquery in a view that now does not break parsing. If I remember correctly, the difference in rows for you is because you forgot to change `\\d` to `\d` in the regex.
2. **performance**: as previously advertised, the json is not as fast as the previous, optimized regex, that is currently on master. However, this is not a performance degradation, because the big picture is what counts. old master -> json PR is still an improvement. This is kind of similar to the introduction of prepared statements: Even though the `set_config` was a bit slower, in total this is still a nice performance improvement.
3. **relative performance**: Why is the relative execution time ~ 1.3x for me and 2.3x for you, when comparing master and json? My understanding is that the numbers we see are a sum of two components: I call one of them "base" and the other "regex". That might not reflect what is really happening in those two parts, but the conclusion should still hold: We do have different machines and my CPU is quite a bit faster (Ryzen 3950X). My assumption is that the base component of the query (select parsing, IO, ...) does not make that much of a difference between our machines, but the whole regex parsing does differ a lot. I guess this is a lot more dependent on CPU. So the increased complexity in replacing/regexing to the json does not slow me down as much as it does you. I would be really interested in seeing your numbers for **old master** as well though. I really hope they are higher than what you get for the json - so that the "old master -> json" = "improving performance" still holds :).

Of course there could also be other things influencing performance. Which PG version are you testing on?

> > Using the big test schema, I get 1 more available column to embed on and the following timings for n=100:
> > query  | plan [ms] | execution [ms]
> > master | 1.71      | 681.5
> > json   | 2.67      | 574.2
> 
> Bummer, I'm seeing that the new query is about x2 slower than the old one. Same setup with the `timeit` function(n=100):
> query 	plan(ms) 	execution(ms)
> master 	3.38227 	618.11523
> json 	3.52706 	1450.29681
> 
> @wolfgangwalther Have you tested the query again recently? What could be wrong?

Just to make that clear again: **master** is not the same between what I posted back then and what you posted here. Back then master was without #1625, your master is with that included. So it is **expected** that this PR is slower than current master.

wolfgangwalther commented:
> > I get 1 more available column to embed
> 
> I get 10 more records on the big schema when running the new query(578 on the old, and 588 on the new), perhaps #1625 changed the perf here?

In case it get's lost in the big text: you need to change `\\d` to `\d` in the regex, if I remember correctly.

steve-chavez commented:
> because you forgot to change \\d to \d in the regex

Argh, forgot about that one.

> I would be really interested in seeing your numbers for old master as well though

Sure. I just ran your https://pastebin.com/9CYCzCxc. I had to delete the count though, it erred with `column "last_count" does not exist`.

```
┌────────────┬──────┬───────────┐
│   query    │ plan │ execution │
├────────────┼──────┼───────────┤
│ old master │ 1.13 │     824.0 │
│ master     │ 3.44 │     627.0 │
│ json PR    │ 3.44 │    1452.8 │
└────────────┴──────┴───────────┘
```

> Which PG version are you testing on?

It's on 11.5.

steve-chavez commented:
Ran the same test on a newer laptop(system76 oryx pro), pg 12.5:

```
   query    | plan | execution
------------+------+-----------
 old master | 0.57 |     288.7
 master     | 1.06 |     173.6
 json PR    | 1.02 |     258.2
```

<details>
<summary>Oryx pro lscpu(16 CPUs)</summary>
<br>
<pre>
$ lscpu 
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Byte Order:                      Little Endian
Address sizes:                   39 bits physical, 48 bits virtual
CPU(s):                          16
On-line CPU(s) list:             0-15
Thread(s) per core:              2
Core(s) per socket:              8
Socket(s):                       1
NUMA node(s):                    1
Vendor ID:                       GenuineIntel
CPU family:                      6
Model:                           165
Model name:                      Intel(R) Core(TM) i7-10875H CPU @ 2.30GHz
Stepping:                        2
CPU MHz:                         1905.241
CPU max MHz:                     5100.0000
CPU min MHz:                     800.0000
BogoMIPS:                        4599.93
Virtualization:                  VT-x
L1d cache:                       256 KiB
L1i cache:                       256 KiB
L2 cache:                        2 MiB
L3 cache:                        16 MiB
NUMA node0 CPU(s):               0-15
Vulnerability Itlb multihit:     KVM: Mitigation: VMX disabled
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and
                                 seccomp
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sani
                                 tization
Vulnerability Spectre v2:        Mitigation; Enhanced IBRS, IBPB conditional, RSB filling
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected
</pre>
</details>

wolfgangwalther commented:
> Sure. I just ran your https://pastebin.com/9CYCzCxc. I had to delete the count though, it erred with `column "last_count" does not exist`.

Ah, yes - that was my upgraded timeit function, sorry. It only adds the count, though.


> Oryx pro lscpu(16 CPUs)
> 
> @wolfgangwalther Seems this one is faster than your Ryzen stuck_out_tongue. I'll try pg 12.5 on the old toshiba just in case.

Whooooot? I gotta get a new CPU!! :D

Small correction, though: I just realized that I was actually running **PG 13**, sorry. But the numbers of 12.4 were exactly the same back then, so I don't think there is a difference.

> Yeah, this one is an old toshiba laptop, running pg 11.5. Here's the lscpu output:
> Toshiba lscpu(4 CPUs)

I do think, however that there is a difference between 11.5 and 12/13? Or maybe it is the machine.

---

So.. let's not think too much about why your laptop is faster than my Ryzen... terrible... but focus on the fact, that they both show the same pattern, kind of.

But the Toshiba / PG 11 shows something entirely different. If you could get some numbers on PG12/13 on that laptop, that would be cool. I'm going to run the test on PG 11 on my machine now. That should give us a good idea of whats happening.

wolfgangwalther commented:
I went back and did the test for some more versions. I only ran `n=10`, but my experience tells me that the numbers are quite consistent with that as well. I removed the plan time and the count and am focusing on the meat, the execution time:

| query | pg10 | pg11 | pg12 | pg13 |
| -- | --: | --: | --: | --: |
| old master | 311.6 | 295.7 | 458.5 | 479.2 |
| master | 230.3 | 222.2 | 253.0 | 263.0 |
| json PR | 523.9 | 504.8 | 319.5 | 336.6 |

PG just got slower on the regular regex and faster on the JSON stuff between 11 and 12. I think the CPU stuff still plays a role, though, as one can tell from your laptop data. Not only is it faster, but the relations between those queries are also different.

One thing to note here: The json query has been optimized while running PG12. So maybe it could be optimized to run faster with the older versions as well. But I don't think we need that. We should mostly be concerned about performance with newer PG versions, I think. Especially on this query, which is "just" the schema cache and startup time.

steve-chavez commented:
Can confirm the pattern on the old toshiba with pg 12.4:

| query      | plan | execution  |
| ---------- | ---- | ---------- |
| old master | 2.34 | 1233.7     |
| master     | 4.70 | 723.4      |
| json PR    | 4.99 | 1029.4     |

> We should mostly be concerned about performance with newer PG versions, I think.

Yeah, totally agree. But should we document this change somehow? Like "#1632, expect a faster startup time for pg > 12, a bit slower for pg < 12"? Might get some users to upgrade pg as well(if the change is noticeable for them).

wolfgangwalther commented:
> Yeah, totally agree. But should we document this change somehow? Like "#1632, expect a faster startup time for pg > 12, a bit slower for pg < 12"? Might get some users to upgrade pg as well(if the change is noticeable for them).

Yes, that makes sense!

steve-chavez commented:
So this one only needs some comments added.

Besides that, it should be good to go :+1:

wolfgangwalther commented:
Ready from my side, will just wait for CI before merging.
